modules/grpc/pubsub/filterx/object-pubsub-message.cpp

```diff
@@ -0,0 +1,352 @@
+}
+
+FilterXObject *
+_filterx_pubsub_message_clone(FilterXObject *s)
+{
+  FilterXPubSubMessage *self = (FilterXPubSubMessage *) s;
+
+  FilterXPubSubMessage *clone = g_new0(FilterXPubSubMessage, 1);
+  _init_instance(clone);
+
+  try
+    {
+      clone->cpp = new Message(*self->cpp, clone);
+    }
+  catch (const std::runtime_error &)
+    {
+      g_assert_not_reached();
+    }
+
+  return &clone->super;
+}
+
+gboolean
+_get_repr(FilterXObject *obj, std::string &str, GString *repr)
+{
+  const gchar *buf = NULL;
+  gsize len;
+  if (filterx_object_extract_string_ref(obj, &buf, &len))
+    {
+      str = std::string(buf, len);
+    }
+  else
+    {
+      g_string_truncate(repr, 0);
+      if (!filterx_object_repr(obj, repr))
+        return FALSE;
+      str = std::string(repr->str, repr->len);
+    }
+  return TRUE;
+}
+
+gboolean
+_build_map(FilterXObject *key, FilterXObject *val, gpointer user_data)
+{
+
+  auto *msg = static_cast<syslogng::grpc::pubsub::filterx::Message *>(static_cast<gpointer *>(user_data)[0]);
+  GString *buf = static_cast<GString *>(static_cast<gpointer *>(user_data)[1]);
+
+  std::string key_cpp;
+  std::string val_cpp;
+
+  if (!_get_repr(key, key_cpp, buf))
+    return FALSE;
+  if (!_get_repr(val, val_cpp, buf))
+    return FALSE;
+
+  return msg->set_attribute(key_cpp, val_cpp);
+}
+
+FilterXObject *
+filterx_pubsub_message_new_from_args(FilterXExpr *s, FilterXObject *args[], gsize args_len)
+{
+  FilterXPubSubMessage *self = g_new0(FilterXPubSubMessage, 1);
+  _init_instance(self);
+
+  try
+    {
+      if (!args || args_len == 0)
+        {
+          self->cpp = new Message(self);
+        }
+      else if (args_len == 2)
+        {
+          FilterXObject *data = args[0];
+          FilterXObject *data_arg = filterx_ref_unwrap_ro(data);
+          FilterXObject *attributes = args[1];
+          FilterXObject *attributes_arg = filterx_ref_unwrap_ro(attributes);
+          if (filterx_object_is_type(attributes_arg, &FILTERX_TYPE_NAME(dict)))
+            {
+              self->cpp = new Message(self);
+              std::string data_cpp;
+              gsize len = 0;
+              if (filterx_object_is_type(data_arg, &FILTERX_TYPE_NAME(string)))
+                {
+                  const gchar *data_str = filterx_string_get_value_ref(data_arg, &len);
+                  data_cpp = std::string(data_str, len);
+                }
+              else if (filterx_object_is_type(data_arg, &FILTERX_TYPE_NAME(bytes)))
+                {
+                  const gchar *data_str = filterx_bytes_get_value_ref(data_arg, &len);
+                  data_cpp = std::string(data_str, len);
+                }
+              else if (filterx_object_is_type(data_arg, &FILTERX_TYPE_NAME(protobuf)))
+                {
+                  const gchar *data_str = filterx_protobuf_get_value_ref(data_arg, &len);
+                  data_cpp = std::string(data_str, len);
+                }
```

Comment:
Sorry, here also I forgot about `message_value`s. I think we should use the extractor here, but the good news is that it will make the code a lot more simpler, as you can call each `extract()` in one `if` condition.